migration: Update message `edit_history` fields.

laurynmm opened the pull request:
Updates the database entries for `message.edit_history` (and archived messages as well) for the following:

- Replaces edit history logs containing legacy `prev_subject` field with `prev_topic` field.
- All edit history logs with:
  - `prev_topic` field now also have `topic` field.
  - `prev_stream` field now also have `new_stream` field.

Makes database migrations for edit history fields re: #21076.

- Does not change field type to JSONB field, [re: this comment](https://github.com/zulip/zulip/issues/21076#issuecomment-1034264914)
- First version of WIP does not include API documentation updates / changes (DONE).

**Testing plan:** 
- Wrote unit test in `test_migrations.py` for database migrations / changes to  message edit history field.
- Did not write a similar suite for archived messages, but manually tested via interacting with the db shell in the dev environment.
- Also, did manual testing in the dev environment for message edits.
- Updated existing js-node and back end tests for changes.
- Noticed that the back end tests for message `edit_history` data structure do not check for the already present `prev_stream` (and therefore do not have the newly present `new_stream` field, so adding that to `test_edit_cases` in `test_message_edit.py` (DONE).

mateuszmandera reviewed (commented):
In general the approach looks great to me - I've posted a bunch of notes.

I'm trying to think about the details of how this will be deployed:

The migration iterates over all the Messages, so probably will take a while to complete - meaning we'll need it to be running while the server is live. (This is less of a concern for self-hosted deployments, but a real one for Zulip Cloud)

So, I think we'll want to deploy the commit with that migration together with a change to the server code that, when a message is edited, adds the relevant entry to both `edit_history_update_fields` and `edit_history` columns - so that message edits happening during the migration's runtime don't cause an inconsistent state again.

@timabbott Does that sound right?

mateuszmandera commented on zerver/migrations/0378_message_edit_history_update_fields.py:
```diff
@@ -0,0 +1,81 @@
+    for message in messages:
+        edit_history = orjson.loads(message.edit_history)
+        topic = message.subject
+        stream = message.recipient_id
```
Hmm, in the `do_update_message` codepath we do `edit_history_event["prev_stream"] = stream_being_edited.id`, so looks like we want the `Stream.id` here. You can get that by `Recient.objects.get(id=message.recipient_id, type=Recipient.STREAM).type_id`

laurynmm replied:
Thank you for clarifying the relationship here. After making my revision to the migration logic, I added a double check for the stream id in my migration test for a stream only edit.

mateuszmandera commented:
I mentioned it in a comment, but worth a separate note since it requires some attention - I'm not familiar with the "update message" codepath off the top of my head, so we'll want to investigate about whether we need any special distinctions in handling stream messages vs private messages.

laurynmm commented on zerver/migrations/0378_message_edit_history_update_fields.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def backfill_message_edit_history(first_id: int, last_id: int, Message: Model) -> None:
+    messages = Message.objects.filter(edit_history__isnull=False, id__range=(first_id, last_id))
```
I added this in my revision so that now the four columns that are fetched from the Message table are: "subject", "recipient", "edit_history", "edit_history_update_fields".

laurynmm commented on zerver/migrations/0378_message_edit_history_update_fields.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def backfill_message_edit_history(first_id: int, last_id: int, Message: Model) -> None:
+    messages = Message.objects.filter(edit_history__isnull=False, id__range=(first_id, last_id))
```
I also included this in my revision of the migration - so it has both the `select_for_update()` and `only()`.

laurynmm commented:
@mateuszmandera - Your review pass was very helpful! I feel like I've learned so much working on this migration this past week. I've gone through and implemented your suggestions / comments. I think it's ready for another pass.

**Re: the server being live during the migration**:
The logic I added for continuing to add edit history updates during the migration (since the server will be live) probably deserves a solid double-check. The affected files are `actions.py` and `topic.py` in the `migration: Update message edit_history entries` commit. I added some comments to articulate my thought process there, which seemed okay since it all gets deleted in a later commit.

**Re: the concern about private messages logic**:
As I went through the code and tests, I'm pretty sure the "update message" code path is already limiting topic and stream edits to only be possible for stream messages. Private messages and Huddles (group pms, right?) should already only be allowed content edits, which we then only copy in the migration.

I did add a check in the migration (0378) for the message recipient type so that we explicitly only go through the process of updating / renaming the fields for stream messages.

@timabbott - I've added the documentation commit to this latest WIP version, if you want to look it over. Here's the [HTML api docs diff](https://pastebin.com/va4USiNh). I didn't include the `/zulip-outgoing-webhooks` endpoint in the API changelog entry even though it does get updated by the additions to the OpenAPI doc.

laurynmm commented:
One other question / thought I'd like clarified is about whether we want to also update the field type for the `message.edit_history` from `string` to `JSONB`. The current structure of this PR should allow for making that change easily. But I haven't really focused on it due to [this comment from the issue](https://github.com/zulip/zulip/issues/21076#issuecomment-1034264914).

timabbott commented:
I think it's probably simpler to leave the JSONB migration for a follow-up project.

laurynmm commented:
I'm going to go ahead and move this from WIP draft to ready for review. I've updated the API documentation to read better (see screenshots below). Here's a link to the updated [HTML diff](https://pastebin.com/rcJAGGV0).

I also added two more tests to the data migration commit (`Update message edit_history entries`). Those tests will only pass for that commit because they require both columns to exist, so I mark them with `#no coverage` in the subsequent commits. My goal was have written tests for the logic I added for having both columns update for any edits that happen during the migration. I had already done a bunch of manual testing in the dev environment.

@mateuszmandera and @timabbott - I'm not sure on the timeline for this re: the release goals, but I definitely want to make sure that everything looks solid since this is my first complex migration project. Let me know if there's more I can do at this point. Thanks!

**API documentation screenshots**:
1. Changelog:
![Screenshot from 2022-02-25 15-12-46](https://user-images.githubusercontent.com/63245456/155755275-c5814ff4-e0b1-4bbc-9504-d72a888d11b9.png)

3. Message object's edit history object (appears in multiple locations in docs):
![Screenshot from 2022-02-25 15-20-20](https://user-images.githubusercontent.com/63245456/155755253-a49ce770-2993-41e4-ac4a-e33a20eb7a47.png)

4. Get message history endpoint return value:
![Screenshot from 2022-02-25 15-13-29](https://user-images.githubusercontent.com/63245456/155755238-531e3fbd-038e-45d2-8ac9-6a883aa7b681.png)

laurynmm commented:
@timabbott & @mateuszmandera  - These most recent updates include the `JSONField` migration in the `Message` model. 

**JSONField notes**:
The updates for that change to the rest of the code base are in a separate commit for now. I thought that might help with reviewing and, in the case that we decide against it for some reason, would be easier for me to undo that addition to this migration. 

I used `TypedDict` instead of a more generic `List[object]` or `List[Dict[str, Any]` in the model type definition. I also defined another `TypedDict` for the dictionary that's returned by the [`/get-message-history`](https://zulip.com/api/get-message-history) endpoint (which was previously modifying the old data structure of `edit_history` list).

**Live migration notes**:
Also, the code processing edit events during the live migration have been updated so there's less temporary code duplication in `actions.py` for the `edit_history_event`. This required more complicated logic in `topic.py` for those cases.

I chose to temporarily save data with the new fields (`topic`, `prev_topic` and `stream`) in the old `edit_history` column since it's deleted in the following commit and since those fields temporary existence shouldn't impact the back and front end code that uses the `edit_history`. I focused on ensuring that the data in the new `edit_history_update_fields` column was correct.

But I could go in and make sure that `topic.py` is creating edit history entries that are correct for both columns for message edit events that come in during the migration if you all think that's more prudent.

**Documentation notes**:
I didn't change any of the documentation updates, so the screenshots and HTML diff link above are still accurate.

timabbott commented on zerver/models.py:
```diff
@@ -2704,6 +2705,7 @@ class AbstractMessage(models.Model):
     # A JSON-encoded list of objects describing any past edits to this
     # message, oldest first.
     edit_history: Optional[str] = models.TextField(null=True)
+    edit_history_update_fields: Optional[List[EditHistoryEvent]] = models.JSONField(null=True)
```
Why "_update_fields"?  I would probably just call this `edit_history_jsonb`.

timabbott commented on zerver/tests/test_migrations.py:
```diff
@@ -344,8 +344,8 @@ def test_multiple_stream_and_topic_edits(self) -> None:
-class MessageEditAfterMigrationWithEditHistory(MigrationsTestCase):
-    __unittest_skip__ = False
+class MessageEditAfterMigrationWithEditHistory(MigrationsTestCase):  # nocoverage
+    __unittest_skip__ = True
```
Should we just delete these tests in this commit; we don't necessarily keep any but the example and there's little point in keeping tests we won't run?

timabbott commented on zerver/migrations/0379_message_edit_history_remove_field.py:
```diff
@@ -0,0 +1,19 @@
+            model_name="message",
+            name="edit_history",
+        ),
+    ]
```
This migration needs to be after the commit that switches to reading from the new JSON field -- generally when deploying a complex migration like this, we'll want to deploy individual commits one-by-one, and we can't delete these fields if they are still being accessed.  

(Usually, removing the old fields will be the last commit).

timabbott commented:
Thanks for all your work on this @laurynmm!  

I opened #21282 which takes most of the code from this and restructures it into a commit series that I think it's realistic we can merge. I flagged a few specific issues above with the ordering of changes here; but then realized that doing the sort of giant combined API/data/database format change we've ended up with here is likely to be prohibitively difficult to safely deploy to production.

Sorry I didn't recognize that we should change direction earlier in development! It's probably going to be fairly annoying to clean up the #21282 version, but I feel good about that being a deployable path forward for this project.

timabbott commented:
I think probably it makes sense to close this and just make a new pull request when we're ready to work on the JSONB format migration using this as a reference, but I'll leave that up to you.

laurynmm commented:
Yeah, I totally agree.
